File: bl/questions.py
"""
bl/questions.py — Question bank I/O.

Parses questions.md and reads/writes question status via results.tsv.
"""

import re
from pathlib import Path

from bl.config import cfg

# C-30: tags that require live evidence vs. static analysis only
_BEHAVIORAL_TAGS = frozenset(
    ("performance", "correctness", "agent", "http", "benchmark")
)
_CODE_AUDIT_TAGS = frozenset(("quality", "static", "code-audit"))


def parse_questions() -> list[dict]:
    """Parse questions.md and return list of question dicts."""
    text = cfg.questions_md.read_text(encoding="utf-8")

    block_pattern = re.compile(
        r"^## ([\w][\w.-]*) \[(\w+)\] (.+?)$",  # F4.3: accept BL 2.0 IDs (D1, F2.1, A4...) not just Q-prefix
        re.MULTILINE,
    )
    field_pattern = re.compile(
        r"^\*\*(Mode|Target|Hypothesis|Test|Verdict threshold|Agent|Finding|Source|Operational Mode|Resume After)\*\*:\s*(.+?)(?=\n\*\*|\Z)",
        re.MULTILINE | re.DOTALL,
    )

    matches = list(block_pattern.finditer(text))
    questions = []

    for i, m in enumerate(matches):
        qid = m.group(1)
        tag_raw = m.group(2)
        mode_raw = tag_raw.lower()
        title = m.group(3).strip()

        start = m.end()
        end = matches[i + 1].start() if i + 1 < len(matches) else len(text)
        body = text[start:end]

        fields = {}
        for fm in field_pattern.finditer(body):
            fields[fm.group(1).lower().replace(" ", "_")] = fm.group(2).strip()

        # C-30: derive question_type from header tag or explicit **Type** field
        explicit_type = fields.get("type", "").lower()
        tag_lower = tag_raw.lower()
        if explicit_type in ("behavioral", "code_audit"):
            question_type = explicit_type
        elif tag_lower in _CODE_AUDIT_TAGS:
            question_type = "code_audit"
        elif tag_lower in _BEHAVIORAL_TAGS:
            question_type = "behavioral"
        else:
            question_type = "behavioral"  # default

        status = get_question_status(qid)

        questions.append(
            {
                "id": qid,
                "mode": fields.get(
                    "mode", mode_raw
                ),  # F5.1: body **Mode** field takes priority over bracket tag
                "title": title,
                "status": status,
                "question_type": question_type,
                "target": fields.get("target", ""),
                "hypothesis": fields.get("hypothesis", ""),
                "test": fields.get("test", ""),
                "verdict_threshold": fields.get("verdict_threshold", ""),
                "agent_name": fields.get("agent", "").strip(),
                "finding": fields.get("finding", "").strip(),
                "source": fields.get("source", "").strip(),
                "operational_mode": fields.get("operational_mode", "diagnose"),
                "resume_after": fields.get("resume_after", "").strip(),
            }
        )

    return questions


def load_questions(path: str) -> list[dict]:
    """Parse questions.md from an arbitrary path, using results.tsv in the same directory."""
    questions_path = Path(path)
    results_path = questions_path.parent / "results.tsv"
    text = questions_path.read_text(encoding="utf-8")

    block_pattern = re.compile(
        r"^## ([\w][\w.-]*) \[(\w+)\] (.+?)$",
        re.MULTILINE,
    )
    field_pattern = re.compile(
        r"^\*\*(Mode|Target|Hypothesis|Test|Verdict threshold|Agent|Finding|Source|Operational Mode|Resume After)\*\*:\s*(.+?)(?=\n\*\*|\Z)",
        re.MULTILINE | re.DOTALL,
    )

    def _status(qid: str) -> str:
        if not results_path.exists():
            return "PENDING"
        for line in results_path.read_text(
            encoding="utf-8", errors="replace"
        ).splitlines():
            parts = line.split("\t")
            if parts and parts[0] == qid:
                return parts[1].strip() if len(parts) > 1 else "PENDING"
        return "PENDING"

    matches = list(block_pattern.finditer(text))
    questions = []

    for i, m in enumerate(matches):
        qid = m.group(1)
        tag_raw = m.group(2)
        mode_raw = tag_raw.lower()
        title = m.group(3).strip()

        start = m.end()
        end = matches[i + 1].start() if i + 1 < len(matches) else len(text)
        body = text[start:end]

        fields = {}
        for fm in field_pattern.finditer(body):
            fields[fm.group(1).lower().replace(" ", "_")] = fm.group(2).strip()

        tag_lower = tag_raw.lower()
        explicit_type = fields.get("type", "").lower()
        if explicit_type in ("behavioral", "code_audit"):
            question_type = explicit_type
        elif tag_lower in _CODE_AUDIT_TAGS:
            question_type = "code_audit"
        elif tag_lower in _BEHAVIORAL_TAGS:
            question_type = "behavioral"
        else:
            question_type = "behavioral"

        questions.append(
            {
                "id": qid,
                "mode": fields.get("mode", mode_raw),
                "title": title,
                "status": _status(qid),
                "question_type": question_type,
                "target": fields.get("target", ""),
                "hypothesis": fields.get("hypothesis", ""),
                "test": fields.get("test", ""),
                "verdict_threshold": fields.get("verdict_threshold", ""),
                "agent_name": fields.get("agent", "").strip(),
                "finding": fields.get("finding", "").strip(),
                "source": fields.get("source", "").strip(),
                "operational_mode": fields.get("operational_mode", "diagnose"),
                "resume_after": fields.get("resume_after", "").strip(),
            }
        )

    return questions


def get_question_status(qid: str) -> str:
    """Read current verdict from results.tsv. Returns 'PENDING' if not found."""
    if not cfg.results_tsv.exists():
        return "PENDING"
    for line in cfg.results_tsv.read_text(
        encoding="utf-8", errors="replace"
    ).splitlines():
        parts = line.split("\t")
        if parts and parts[0] == qid:
            return parts[1].strip() if len(parts) > 1 else "PENDING"
    return "PENDING"


_PARKED_STATUSES = frozenset(
    (
        "DIAGNOSIS_COMPLETE",
        "PENDING_EXTERNAL",
        "DEPLOYMENT_BLOCKED",  # Q6.3: diagnosed but awaiting deployment; re-check suppressed
        "DONE",
        "INCONCLUSIVE",
        "FIXED",
        "FIX_FAILED",
        "COMPLIANT",
        "NON_COMPLIANT",
        "CALIBRATED",
        "BLOCKED",
        "HEAL_EXHAUSTED",  # F-mid.1: exhausted heal loop — human intervention required
    )
)

# Question IDs referenced in resume_after fields (e.g. "Q6.3" or "after Q6.3 is DONE")
# are matched by this pattern.
import re as _re

_QREF_PATTERN = _re.compile(r"\b([A-Za-z]\d+(?:\.\d+)?)\b")


def get_next_pending(questions: list[dict]) -> dict | None:
    """Return the first PENDING question, skipping parked/terminal statuses.

    resume_after supports two formats:
      - ISO-8601 datetime string: skip until that time has passed
      - Question-ID reference (e.g. "Q6.3" or "after Q6.3 is DONE"):
        skip until the referenced question reaches a resolved status
    """
    from datetime import datetime, timezone

    # Statuses that count as "resolved" for resume_after question-ID references.
    _RESOLVED = frozenset(
        (
            "DONE",
            "HEALTHY",
            "WARNING",
            "FAILURE",
            "INCONCLUSIVE",
            "FIXED",
            "FIX_FAILED",
            "COMPLIANT",
            "NON_COMPLIANT",
            "CALIBRATED",
            "HEAL_EXHAUSTED",
        )
    )

    now = datetime.now(timezone.utc)
    for q in questions:
        if q["status"] in _PARKED_STATUSES:
            continue
        if q["status"] != "PENDING":
            continue
        resume_after = q.get("resume_after", "")
        if resume_after:
            # Try ISO-8601 datetime first
            try:
                gate = datetime.fromisoformat(resume_after.replace("Z", "+00:00"))
                if now < gate:
                    continue
            except ValueError:
                # Try question-ID reference: "Q6.3", "after Q6.3 is DONE", "DONE:Q6.3"
                ref_match = _QREF_PATTERN.search(resume_after)
                if ref_match:
                    ref_qid = ref_match.group(1).upper()
                    ref_q = get_question_by_id(questions, ref_qid)
                    if ref_q and ref_q["status"] not in _RESOLVED:
                        continue  # referenced question not yet resolved
        return q
    return None


def get_question_by_id(questions: list[dict], qid: str) -> dict | None:
    """Return the question with the given ID, or None."""
    for q in questions:
        if q["id"] == qid:
            return q
    return None


def sync_status_from_results() -> int:
    """
    Reconcile questions.md Status fields against results.tsv.
    Returns count of questions updated.
    """
    if not cfg.results_tsv.exists() or not cfg.questions_md.exists():
        return 0

    # Build map of qid → new status from results.tsv
    done_ids: dict[str, str] = {}
    for line in cfg.results_tsv.read_text(
        encoding="utf-8", errors="replace"
    ).splitlines():
        parts = line.split("\t")
        if len(parts) >= 2 and parts[0] not in ("question_id", ""):
            verdict = parts[1].strip()
            _TERMINAL_VERDICTS = frozenset(
                {
                    "HEALTHY",
                    "WARNING",
                    "FAILURE",
                    "INCONCLUSIVE",
                    "DIAGNOSIS_COMPLETE",
                    "PENDING_EXTERNAL",
                    "FIXED",
                    "FIX_FAILED",
                    "PROMISING",
                    "WEAK",
                    "BLOCKED",
                    "CALIBRATED",
                    "UNCALIBRATED",
                    "NOT_MEASURABLE",
                    "COMPLIANT",
                    "NON_COMPLIANT",
                    "PARTIAL",
                    "NOT_APPLICABLE",
                    "IMPROVEMENT",
                    "REGRESSION",
                    "IMMINENT",
                    "PROBABLE",
                    "POSSIBLE",
                    "UNLIKELY",
                    "OK",
                    "DEGRADED",
                    "ALERT",
                    "UNKNOWN",
                    "SUBJECTIVE",
                    "DEGRADED_TRENDING",
                    "DEPLOYMENT_BLOCKED",  # Q6.3: preserve — human must confirm deployment before unblocking
                    "HEAL_EXHAUSTED",  # F-mid.1: exhausted heal loop — preserve status
                }
            )
            if verdict in _TERMINAL_VERDICTS:
                done_ids[parts[0]] = (
                    verdict
                    if verdict
                    in (
                        "INCONCLUSIVE",
                        "DIAGNOSIS_COMPLETE",
                        "PENDING_EXTERNAL",
                        "DEPLOYMENT_BLOCKED",  # Q6.3: preserve — awaiting deployment
                        "FIXED",
                        "FIX_FAILED",
                        "BLOCKED",
                        # F11.2: preserve failure/violation verdicts for human visibility (matches F8.2 in findings.py)
                        "FAILURE",
                        "NON_COMPLIANT",
                        "WARNING",
                        "REGRESSION",
                        "ALERT",
                        "HEAL_EXHAUSTED",  # F-mid.1: preserve exhausted-heal status
                    )
                    else "DONE"
                )

    # Update questions.md
    text = cfg.questions_md.read_text(encoding="utf-8")
    updated = 0
    for qid, new_status in done_ids.items():
        block_start = text.find(f"## {qid} [")
        if block_start == -1:
            block_start = text.find(f"## {qid}\n")
        if block_start == -1:
            continue
        next_block = text.find("\n## ", block_start + 1)
        block_end = next_block if next_block != -1 else len(text)
        block = text[block_start:block_end]
        if "**Status**: PENDING" in block:
            new_block = block.replace(
                "**Status**: PENDING", f"**Status**: {new_status}", 1
            )
            text = text[:block_start] + new_block + text[block_end:]
            updated += 1

    if updated:
        cfg.questions_md.write_text(text, encoding="utf-8")
    return updated

File: masonry/src/engine/questions.js
'use strict';
// engine/questions.js — Question bank I/O.
//
// Port of bl/questions.py to Node.js. Parses questions.md and
// reads/writes question status via results.tsv.

const fs = require('fs');
const path = require('path');
const { cfg } = require('./config');

// C-30: tags that require live evidence vs. static analysis only
const BEHAVIORAL_TAGS = new Set([
  'performance', 'correctness', 'agent', 'http', 'benchmark',
]);
const CODE_AUDIT_TAGS = new Set([
  'quality', 'static', 'code-audit',
]);

const BLOCK_PATTERN = /^## ([\w][\w.-]*) \[(\w+)\] (.+?)$/gm;
const FIELD_PATTERN = /^\*\*(Mode|Target|Hypothesis|Test|Verdict threshold|Agent|Finding|Source|Operational Mode|Resume After)\*\*:\s*(.+?)(?=\n\*\*|$)/gms;

// Statuses that are parked / terminal — getNextPending skips these
const PARKED_STATUSES = new Set([
  'DIAGNOSIS_COMPLETE', 'PENDING_EXTERNAL', 'DEPLOYMENT_BLOCKED',
  'DONE', 'INCONCLUSIVE', 'FIXED', 'FIX_FAILED',
  'COMPLIANT', 'NON_COMPLIANT', 'CALIBRATED', 'BLOCKED', 'HEAL_EXHAUSTED',
]);

// Statuses that count as "resolved" for resume_after question-ID references
const RESOLVED_STATUSES = new Set([
  'DONE', 'HEALTHY', 'WARNING', 'FAILURE', 'INCONCLUSIVE',
  'FIXED', 'FIX_FAILED', 'COMPLIANT', 'NON_COMPLIANT',
  'CALIBRATED', 'HEAL_EXHAUSTED',
]);

// Terminal verdicts from results.tsv that trigger status sync
const TERMINAL_VERDICTS = new Set([
  'HEALTHY', 'WARNING', 'FAILURE', 'INCONCLUSIVE',
  'DIAGNOSIS_COMPLETE', 'PENDING_EXTERNAL', 'FIXED', 'FIX_FAILED',
  'PROMISING', 'WEAK', 'BLOCKED', 'CALIBRATED', 'UNCALIBRATED',
  'NOT_MEASURABLE', 'COMPLIANT', 'NON_COMPLIANT', 'PARTIAL',
  'NOT_APPLICABLE', 'IMPROVEMENT', 'REGRESSION', 'IMMINENT',
  'PROBABLE', 'POSSIBLE', 'UNLIKELY', 'OK', 'DEGRADED', 'ALERT',
  'UNKNOWN', 'SUBJECTIVE', 'DEGRADED_TRENDING',
  'DEPLOYMENT_BLOCKED', 'HEAL_EXHAUSTED',
]);

// Verdicts preserved as-is in questions.md (not collapsed to DONE)
const PRESERVE_VERDICTS = new Set([
  'INCONCLUSIVE', 'DIAGNOSIS_COMPLETE', 'PENDING_EXTERNAL',
  'DEPLOYMENT_BLOCKED', 'FIXED', 'FIX_FAILED', 'BLOCKED',
  'FAILURE', 'NON_COMPLIANT', 'WARNING', 'REGRESSION',
  'ALERT', 'HEAL_EXHAUSTED',
]);

const QREF_PATTERN = /\b([A-Za-z]\d+(?:\.\d+)?)\b/;

// ---------------------------------------------------------------------------
// Status lookup
// ---------------------------------------------------------------------------

/**
 * Read current verdict from results.tsv. Returns 'PENDING' if not found.
 */
function getQuestionStatus(qid) {
  if (!fs.existsSync(cfg.resultsTsv)) return 'PENDING';
  const lines = fs.readFileSync(cfg.resultsTsv, 'utf8').split('\n');
  for (const line of lines) {
    const parts = line.split('\t');
    if (parts[0] === qid) {
      return (parts[1] || '').trim() || 'PENDING';
    }
  }
  return 'PENDING';
}

// ---------------------------------------------------------------------------
// Question type classification
// ---------------------------------------------------------------------------

function _classifyType(tagRaw, fields) {
  const explicitType = (fields.type || '').toLowerCase();
  if (explicitType === 'behavioral' || explicitType === 'code_audit') return explicitType;
  const tagLower = tagRaw.toLowerCase();
  if (CODE_AUDIT_TAGS.has(tagLower)) return 'code_audit';
  if (BEHAVIORAL_TAGS.has(tagLower)) return 'behavioral';
  return 'behavioral';
}

// ---------------------------------------------------------------------------
// Parsing
// ---------------------------------------------------------------------------

function _parseBody(body) {
  const fields = {};
  let fm;
  const re = new RegExp(FIELD_PATTERN.source, FIELD_PATTERN.flags);
  while ((fm = re.exec(body)) !== null) {
    const key = fm[1].toLowerCase().replace(/ /g, '_');
    fields[key] = fm[2].trim();
  }
  return fields;
}

/**
 * Parse questions.md from cfg.questionsMd and return list of question dicts.
 */
function parseQuestions() {
  const text = fs.readFileSync(cfg.questionsMd, 'utf8');
  const re = new RegExp(BLOCK_PATTERN.source, BLOCK_PATTERN.flags);
  const matches = [];
  let bm;
  while ((bm = re.exec(text)) !== null) {
    matches.push({ index: bm.index, end: bm.index + bm[0].length, groups: [bm[1], bm[2], bm[3]] });
  }

  const questions = [];
  for (let i = 0; i < matches.length; i++) {
    const m = matches[i];
    const [qid, tagRaw, titleRaw] = m.groups;
    const title = titleRaw.trim();
    const start = m.end;
    const end = i + 1 < matches.length ? matches[i + 1].index : text.length;
    const body = text.slice(start, end);
    const fields = _parseBody(body);

    questions.push({
      id: qid,
      mode: fields.mode || tagRaw.toLowerCase(),
      title,
      status: getQuestionStatus(qid),
      question_type: _classifyType(tagRaw, fields),
      target: fields.target || '',
      hypothesis: fields.hypothesis || '',
      test: fields.test || '',
      verdict_threshold: fields.verdict_threshold || '',
      agent_name: (fields.agent || '').trim(),
      finding: (fields.finding || '').trim(),
      source: (fields.source || '').trim(),
      operational_mode: fields.operational_mode || 'diagnose',
      resume_after: (fields.resume_after || '').trim(),
    });
  }

  return questions;
}

/**
 * Parse questions.md from an arbitrary path, using results.tsv in the same dir.
 */
function loadQuestions(filePath) {
  const text = fs.readFileSync(filePath, 'utf8');
  const resultsPath = path.join(path.dirname(filePath), 'results.tsv');

  function localStatus(qid) {
    if (!fs.existsSync(resultsPath)) return 'PENDING';
    const lines = fs.readFileSync(resultsPath, 'utf8').split('\n');
    for (const line of lines) {
      const parts = line.split('\t');
      if (parts[0] === qid) return (parts[1] || '').trim() || 'PENDING';
    }
    return 'PENDING';
  }

  const re = new RegExp(BLOCK_PATTERN.source, BLOCK_PATTERN.flags);
  const matches = [];
  let bm;
  while ((bm = re.exec(text)) !== null) {
    matches.push({ index: bm.index, end: bm.index + bm[0].length, groups: [bm[1], bm[2], bm[3]] });
  }

  const questions = [];
  for (let i = 0; i < matches.length; i++) {
    const m = matches[i];
    const [qid, tagRaw, titleRaw] = m.groups;
    const title = titleRaw.trim();
    const start = m.end;
    const end = i + 1 < matches.length ? matches[i + 1].index : text.length;
    const body = text.slice(start, end);
    const fields = _parseBody(body);

    questions.push({
      id: qid,
      mode: fields.mode || tagRaw.toLowerCase(),
      title,
      status: localStatus(qid),
      question_type: _classifyType(tagRaw, fields),
      target: fields.target || '',
      hypothesis: fields.hypothesis || '',
      test: fields.test || '',
      verdict_threshold: fields.verdict_threshold || '',
      agent_name: (fields.agent || '').trim(),
      finding: (fields.finding || '').trim(),
      source: (fields.source || '').trim(),
      operational_mode: fields.operational_mode || 'diagnose',
      resume_after: (fields.resume_after || '').trim(),
    });
  }

  return questions;
}

// ---------------------------------------------------------------------------
// Next-question logic
// ---------------------------------------------------------------------------

/**
 * Return the first PENDING question, skipping parked/terminal statuses.
 * resume_after supports ISO-8601 datetime or question-ID references.
 */
function getNextPending(questions) {
  const now = new Date();

  for (const q of questions) {
    if (PARKED_STATUSES.has(q.status)) continue;
    if (q.status !== 'PENDING') continue;

    const resumeAfter = q.resume_after || '';
    if (resumeAfter) {
      // Try ISO-8601 datetime first
      const gate = new Date(resumeAfter);
      if (!isNaN(gate.getTime()) && _looksLikeDatetime(resumeAfter)) {
        if (now < gate) continue;
      } else {
        // Try question-ID reference
        const refMatch = resumeAfter.match(QREF_PATTERN);
        if (refMatch) {
          const refQid = refMatch[1].toUpperCase();
          const refQ = getQuestionById(questions, refQid);
          if (refQ && !RESOLVED_STATUSES.has(refQ.status)) continue;
        }
      }
    }

    return q;
  }
  return null;
}

/**
 * Heuristic to distinguish ISO datetime strings from question-ID references.
 */
function _looksLikeDatetime(s) {
  return /^\d{4}-\d{2}/.test(s);
}

/**
 * Return the question with the given ID, or null.
 */
function getQuestionById(questions, qid) {
  for (const q of questions) {
    if (q.id === qid) return q;
  }
  return null;
}

// ---------------------------------------------------------------------------
// Status sync
// ---------------------------------------------------------------------------

/**
 * Reconcile questions.md Status fields against results.tsv.
 * Returns count of questions updated.
 */
function syncStatusFromResults() {
  if (!fs.existsSync(cfg.resultsTsv) || !fs.existsSync(cfg.questionsMd)) return 0;

  // Build map of qid → new status from results.tsv
  const doneIds = {};
  const lines = fs.readFileSync(cfg.resultsTsv, 'utf8').split('\n');
  for (const line of lines) {
    const parts = line.split('\t');
    if (parts.length < 2 || parts[0] === 'question_id' || parts[0] === '') continue;
    const verdict = parts[1].trim();
    if (TERMINAL_VERDICTS.has(verdict)) {
      doneIds[parts[0]] = PRESERVE_VERDICTS.has(verdict) ? verdict : 'DONE';
    }
  }

  // Update questions.md
  let text = fs.readFileSync(cfg.questionsMd, 'utf8');
  let updated = 0;

  for (const [qid, newStatus] of Object.entries(doneIds)) {
    let blockStart = text.indexOf(`## ${qid} [`);
    if (blockStart === -1) blockStart = text.indexOf(`## ${qid}\n`);
    if (blockStart === -1) continue;

    const nextBlock = text.indexOf('\n## ', blockStart + 1);
    const blockEnd = nextBlock !== -1 ? nextBlock : text.length;
    const block = text.slice(blockStart, blockEnd);

    if (!block.includes('**Status**: PENDING')) continue;

    const newBlock = block.replace('**Status**: PENDING', `**Status**: ${newStatus}`);
    text = text.slice(0, blockStart) + newBlock + text.slice(blockEnd);
    updated++;
  }

  if (updated) {
    const tmpPath = path.join(path.dirname(cfg.questionsMd), `.questions-${process.pid}.tmp`);
    try {
      fs.writeFileSync(tmpPath, text, 'utf8');
      fs.renameSync(tmpPath, cfg.questionsMd);
    } catch (err) {
      try { fs.unlinkSync(tmpPath); } catch (_) { /* ignore */ }
      throw err;
    }
  }

  return updated;
}

module.exports = {
  parseQuestions,
  loadQuestions,
  getQuestionStatus,
  getNextPending,
  getQuestionById,
  syncStatusFromResults,
  PARKED_STATUSES,
  RESOLVED_STATUSES,
};
